Merge v1.0 into master

Jake-CSUMB opened the pull request.

cmcmichael-csumb committed:
Followed assignment 2 specifications for editing Assignment, Grade, and Enrollment objects. Uncommented OneToMany relationships in Section.java and uncommented/altered Repository SQL queries for Assignment, Grade and Enrollment.

cmcmichael-csumb committed:
Played around with testing the Assignment, Enrollment, and Grade Repositories using the AssignmentController as a testing tool.

Chrismac112 committed:
Merged master into v1.0 branch, applied Andis changes

Chrismac112 committed:
Added throwing errors to the enrollmentController file

jeremiahmcgrath committed:
Update AssignmentController.java

Completed AssignmentController.java's TODOs apart from one on method getStudentAssignments.

Still need to finish finding grade for each assignment using enrollmentId and assignmentId.

jeremiahmcgrath committed:
Fixed dates in AssignmentController.java

Fixed date types handling, removed parsing from util -> sql date types, opting for str - sql.date handling.

Edited HttpStatus errors for readability.

Commented out code for getStudentAssignments method so main program runs without error. This method is still TODO'd.

jeremiahmcgrath committed:
Finalized AssignmentController.java for v1.0 update

Method getStudentAssignments completed.

All TODO's for AssignmentController.java addressed.

All methods for AssignmentController.java tested for functionality with Postman.

Jake-CSUMB committed:
Implement StudentController

Jake-CSUMB committed:
Fix exception in dropCourse()

cmcmichael-csumb commented on src/main/java/com/cst438/controller/AssignmentController.java:
```diff
@@ -3,102 +3,200 @@
+        List<Assignment> assignments = assignmentRepository.findBySectionNoOrderByDueDate(secNo);
+
+        if (assignments.size() == 0) {
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Section " + secNo + " not found");
```
When a Section has no assignments, this prints "Section # not found". We may want to change and or add to that to say something regarding no assignments linked to such section.

cmcmichael-csumb committed:
Removed needless TODO lines

cmcmichael-csumb committed:
Removed extra spaces and added throwing errors on getAssignmentGrades

and1cam commented on src/main/java/com/cst438/controller/AssignmentController.java:
```diff
@@ -3,102 +3,203 @@
+        // Fetch assignments from repository
+        List<Assignment> assignments = assignmentRepository.findBySectionNoOrderByDueDate(secNo);
+
+        if (assignments.size() == 0) {
```
Alternative: `assignments.isEmpty()`

Jake-CSUMB replied:
Should an empty list be an error state? Perhaps the `secNo` could be validated separately, and an empty list returned for a section with no assignments.

jeremiahmcgrath replied:
If statement changed to alternative: `assignments.isEmpty()`

jeremiahmcgrath replied:
Both suggestions implemented.

and1cam commented on src/main/java/com/cst438/controller/AssignmentController.java:
```diff
@@ -3,102 +3,203 @@
-    public AssignmentDTO updateAssignment(@RequestBody AssignmentDTO dto) {
+    public AssignmentDTO updateAssignment(@RequestBody AssignmentDTO assignmentDTO) {
+
+        Assignment a = assignmentRepository.findById(assignmentDTO.id()).orElse(null);
```
Alternative: `public AssignmentDTO updateAssignment(@RequestBody AssignmentDTO assignmentDTO) {
Assignment a = assignmentRepository.findById(assignmentDTO.id()).orElseThrow(() ->
new ResponseStatusException(HttpStatus.NOT_FOUND, "Assignment not found for ID: " + assignmentDTO.id()));
return getAssignmentDTO(assignmentDTO, a);
    }`

jeremiahmcgrath replied:
Suggestion implemented.

cmcmichael-csumb commented on src/main/java/com/cst438/controller/StudentController.java:
```diff
@@ -7,75 +8,151 @@
+            .findEnrollmentBySectionNoAndStudentId(sectionNo, studentId);
+
+        if (existingEnrollment != null) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "studentId is already enrolled in sectionNo");
```
"studentId is already enrolled in sectionNo"

Is this intended, or should it be 
studentId + " is already enrolled in " + sectionNo

Jake-CSUMB replied:
Intentional, as the caller already knows what studentId and sectionNo are.

and1cam commented on src/main/java/com/cst438/controller/EnrollmentController.java:
```diff
@@ -18,27 +17,35 @@
         // Fetch enrollments from repository
         List<Enrollment> enrollments = enrollmentRepository.findEnrollmentsBySectionNoOrderByStudentName(sectionNo);
 
+        if (enrollments.size() == 0) {
```
Alternative: `enrollments.isEmpty()`

Jake-CSUMB replied:
Should an empty list be an error state? Perhaps the `sectionNo` could be validated separately, and an empty list returned for a section with no enrollments.

jeremiahmcgrath replied:
Both suggestions implemented.

and1cam commented on src/main/java/com/cst438/controller/StudentController.java:
```diff
@@ -7,75 +8,151 @@
-           @RequestParam("year") int year,
-           @RequestParam("semester") String semester,
-           @RequestParam("studentId") int studentId) {
+        User user = userRepository.findById(studentId).orElse(null);
```
Alternative: `User user = userRepository.findById(studentId).orElseThrow(() -> 
new ResponseStatusException(HttpStatus.NOT_FOUND, "Unknown studentId"));`

and1cam commented on src/main/java/com/cst438/controller/StudentController.java:
```diff
@@ -7,75 +8,151 @@
 
-        // remove the following line when done.
-        return null;
+        Section section = sectionRepository.findById(sectionNo).orElse(null);
```
Alternative: `Section section = sectionRepository.findById(sectionNo).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Unknown sectionNo"));`

and1cam commented on src/main/java/com/cst438/controller/StudentController.java:
```diff
@@ -7,75 +8,151 @@
+        Enrollment existingEnrollment = enrollmentRepository
+            .findEnrollmentBySectionNoAndStudentId(sectionNo, studentId);
+
+        if (existingEnrollment != null) {
```
Alternative: `if (enrollmentRepository.findEnrollmentBySectionNoAndStudentId(sectionNo, studentId) != null) { throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Student is already enrolled in section"); }`

and1cam commented on src/main/java/com/cst438/controller/StudentController.java:
```diff
@@ -7,75 +8,151 @@
+
+        Enrollment enrollment = enrollmentRepository.findById(enrollmentId).orElse(null);
+
+        if (enrollment == null) {
```
Alternative: `
Enrollment enrollment = enrollmentRepository.findById(enrollmentId).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Unknown enrollmentId"));
if (!enrollment.getUser().getType().equals("STUDENT")) {
throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Only students can drop a course"); }`

and1cam reviewed (approved):
Minor alternatives suggested

Jake-CSUMB committed:
StudentController.java - Resolve code review comments

Jake-CSUMB commented on src/main/java/com/cst438/controller/AssignmentController.java:
```diff
@@ -3,102 +3,203 @@
+        Section s = sectionRepository.findById(assignmentDTO.secId()).orElse(null);
+        if (s == null) {
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Section " + s.getSecId() + " not found");
+        }
```
Alternative:
```java
Section s = sectionRepository.findById(assignmentDTO.secId()).orElseThrow(() ->
    new ResponseStatusException(HttpStatus.NOT_FOUND, "Section " + s.getSecId() + " not found");
```

jeremiahmcgrath replied:
Changed to:
`Section s = sectionRepository.findById(assignmentDTO.secId()).orElseThrow(() ->
            new ResponseStatusException(HttpStatus.NOT_FOUND, "Section " + assignmentDTO.secId() + " not found"));`

Jake-CSUMB commented on src/main/java/com/cst438/controller/AssignmentController.java:
```diff
@@ -107,12 +208,52 @@ public List<AssignmentStudentDTO> getStudentAssignments(
+        // findByStudentIdAndYearAndSemesterOrderByDueDate
+
+        List<Assignment> assignments = assignmentRepository.findByStudentIdAndYearAndSemesterOrderByDueDate(studentId, year, semester);
+        if (assignments == null) {
```
Will this ever be null? Did you mean:
```java
if (assignments.isEmpty()) {
```
Though, an empty list probably isn't an error state, so the check might not be needed at all. Perhaps the existence of the enrollment should be checked separately.

jeremiahmcgrath replied:
Suggested change made.

Jake-CSUMB commented on src/main/java/com/cst438/controller/AssignmentController.java:
```diff
@@ -3,102 +3,203 @@
-        // TODO
+        Assignment a = assignmentRepository.findByAssignmentId(assignmentId);
+        if (a == null) {
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Assignment " + assignmentId + " not found");
```
Deletion should generally be error-free if the resource doesn't exist.

jeremiahmcgrath replied:
Method changed to:
`Assignment a = assignmentRepository.findByAssignmentId(assignmentId);
        assignmentRepository.delete(a);`

Jake-CSUMB reviewed (commented):
Noted a few suggestions.

jeremiahmcgrath committed:
Implemented suggestions made by other reviewers

jeremiahmcgrath reviewed (approved):
All noted changes implemented and all issues address.

nickleus27 commented on src/main/java/com/cst438/controller/AssignmentController.java:
```diff
@@ -3,102 +3,197 @@
+                grades.add(new GradeDTO(g.getGradeId(), u.getName(), u.getEmail(),
+                        a.getTitle(), c.getCourseId(), s.getSecId(), g.getScore()));
+            } else {
+                grades.add(new GradeDTO(new Grade().getGradeId(), u.getName(), u.getEmail(),
+                        a.getTitle(), c.getCourseId(), s.getSecId(), null));
```
I think you want to save the new grade here in the db to auto increment the grade id. See comment on L144.

nickleus27 reviewed (commented):
Hello all. Your code in your PR has very good readability, formatting, and is clean and concise.  All requirements for the assignment are met. I added just one small suggestion. But other than that, this PR definitely LGTM!

jeremiahmcgrath committed:
Update AssignmentController.java

createAssignment fixed to use PK to search for section.

cmcmichael-csumb committed:
Altered the get mapping for grades, a grade with a null score gets saved in the database now.

Jake-CSUMB committed:
Tweak EnrollmentDTO for front end

juanjoseguva commented:
Hello friends! Congratulations on your clean, legible code. No issues jumped out at me, well done! All entities and controllers created, functioning, and readable. Two branches created. Fulfilled all requirements. 50/50

jeremiahmcgrath committed:
updateAssignment, check if assignment title is blank before updating assignment

cmcmichael-csumb committed:
Merge branch 'master' into v1.0

jeremiahmcgrath committed:
Added Unit Tests for adding a new assignment successfully and attempting to add a new assignment with a bad due date

cmcmichael-csumb committed:
Added unit tests for test 4 and 5

jeremiahmcgrath committed:
Added System Test 1 | Instructor adds a new assignment successfully

Jake-CSUMB committed:
Clean up entity classes

Fix database so that Term.termId is autogenerated

Add utility time functions to TestUtils

Jake-CSUMB committed:
Implement AssignmentControllerUnitTests and EnrollmentControllerUnitTests

Jake-CSUMB committed:
Fix a few things in AssignmentControllerUnitTests.java / EnrollmentControllerUnitTests.java

Commit:
Implemented StudentControllerSystemTest and StudentControllerUnitTest

jeremiahmcgrath committed:
Relocated Unit Tests 1 and 2 & System Test 1

cmcmichael-csumb committed:
Implemented System Test 2

cmcmichael-csumb committed:
Merge branch 'v1.0' of github.com:Bitwise-CSUMB/CST438-Assignment2-1 into v1.0

jeremiahmcgrath committed:
Update functionality of AssignmentControllerSystemTest.java, and updated some comments

Jake-CSUMB committed:
Implement EnrollmentControllerSystemTest

Commit:
Updated StudentControllerSystemTest and StudentControllerUnitTest

Commit:
Merge remote-tracking branch 'origin/v1.0' into v1.0

Jake-CSUMB committed:
Various system test fixes

- All system tests now start the backend automatically (and assume it isn't already running)
- All system tests now assert that the frontend is set to the corresponding home page
- Fixed some incorrect tag ids in EnrollmentGradesSystemTest.java

Jake-CSUMB committed:
Only use dummy data for specific unit tests

Jake-CSUMB committed:
Mark file ownership / test numbers

Jake-CSUMB commented on src/test/java/com/cst438/controller/AssignmentControllerSystemTest.java:
```diff
@@ -0,0 +1,151 @@
+    // verify that cst363 is in the list of sections
+    // if it exists, view assignments
+    // Selenium throws NoSuchElementException when the element is not found
+    try {
```
I removed this try-catch / while loop in my system test - not sure what it does.

Jake-CSUMB commented on src/test/java/com/cst438/controller/AssignmentControllerUnitTests.java:
```diff
@@ -0,0 +1,259 @@
+        assertEquals("cst363", a.getSection().getCourse().getCourseId());
+
+        // clean up after test. issue http DELETE request for assignment
+        response = mvc.perform(
```
Since this API isn't part of the test you could directly delete the assignment via `assignmentRepository.deleteById()`. It's fine as-is though.

Jake-CSUMB commented on src/test/java/com/cst438/controller/ControllerUnitTests.java:
```diff
@@ -0,0 +1,151 @@
+        // Confirming that all DTOs were saved successfully.
+        for (int i = 0; i < numDTOs; i++) {
+            Grade g = gradeRepository.findById(newDTOs[i].gradeId()).orElse(null);
+            assertEquals(g.getScore(), newScores[i]);
```
Possible `NullPointerException`; since this is a test that is fine, but maybe a proper assert should be coded.

Jake-CSUMB commented on src/test/java/com/cst438/controller/ControllerUnitTests.java:
```diff
@@ -0,0 +1,151 @@
+        assertEquals(response.getErrorMessage(), "Assignment 55 not found");
+    }
+
+    private static String asJsonString(final Object obj) {
```
Use function in `TestUtils`

Jake-CSUMB commented on src/test/java/com/cst438/controller/ControllerUnitTests.java:
```diff
@@ -0,0 +1,151 @@
+        }
+    }
+
+    private static <T> T  fromJsonString(String str, Class<T> valueType ) {
```
Use function in `TestUtils`

Jake-CSUMB commented on src/test/java/com/cst438/controller/EnrollmentGradesSystemTest.java:
```diff
@@ -0,0 +1,137 @@
+
+        driver.findElement(By.id("saveChanges")).click();
+
+        driver.findElement(By.id("home")).click();
```
Should this also check the database directly?

Jake-CSUMB replied:
No - System Tests don't have direct access to the database.

Jake-CSUMB commented on src/test/java/com/cst438/controller/StudentControllerSystemTest.java:
```diff
@@ -0,0 +1,130 @@
+        }
+
+        // Assertion for section presence in schedule
+        assertTrue(sectionFound, "Enrolled section found");
```
Should this also check the database directly?

Jake-CSUMB replied:
No - System Tests don't have direct access to the database.

Jake-CSUMB commented on src/test/java/com/cst438/controller/StudentControllerUnitTest.java:
```diff
@@ -0,0 +1,166 @@
+
+    }
+
+    private static <T> T  fromJsonString(String str, Class<T> valueType ) {
```
Use function in `TestUtils`

Jake-CSUMB commented on src/test/java/com/cst438/controller/EnrollmentGradesSystemTest.java:
```diff
@@ -0,0 +1,137 @@
+
+        gradeWe.sendKeys(Keys.BACK_SPACE);
+        String newGrade;
+        if (gradeWe.getAttribute("value") == "F")
```
Should use `.equals()`

Jake-CSUMB commented on src/test/java/com/cst438/controller/StudentControllerSystemTest.java:
```diff
@@ -0,0 +1,130 @@
+
+    // System Test #3 - Student enrolls into a section
+    @Test
+    public void systemTestAddCourse() throws Exception {
```
`throws Exception` unneeded

Jake-CSUMB commented on src/test/java/com/cst438/controller/StudentControllerSystemTest.java:
```diff
@@ -0,0 +1,130 @@
+        // Click link to navigate to Enroll in Course page
+        WebElement addCourse = driver.findElement(By.id("addCourse"));
+        addCourse.click();
+        WebElement courseTable = wait.until(ExpectedConditions.visibilityOfElementLocated(By.className("Center")));
```
`courseTable` not used

Jake-CSUMB commented on src/test/java/com/cst438/controller/AssignmentControllerUnitTests.java:
```diff
@@ -0,0 +1,259 @@
+        assertNotEquals(0, result.id());
+        // check other fields of the DTO for expected values
+        assertEquals("Test Assignment 1", result.title());
+        assertEquals("2024-03-01", result.dueDate().toString());
```
`result.dueDate()` already a String - `.toString()` not needed

Jake-CSUMB reviewed (commented):
Noted a few things.

cmcmichael-csumb committed:
Merged assignment controller unit tests

Jake-CSUMB committed:
Purge @SpringBootTest from System Tests

cmcmichael-csumb committed:
Changed EnrollmentGradesSystemTest, I initially misunderstood the requirements

cmcmichael-csumb committed:
Merged origin changes and added a comment in the professors System test

Jake-CSUMB committed:
Minor fixes / tweaks

cmcmichael-csumb committed:
added rabbitmq dependency for assignment 6
